### ISSUE TITLE
 Multi-animal DLC ID Switching Issue

### ISSUE THREAD

**Issue by vasudha-kulkarni** [DeepLabCut/DeepLabCut#2466]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

Hello, I am working on extracting features from a pair of marmosets in the experimental arena from multiple cameras to construct a 3D trajectory of their movements. Each camera captures half of the experimental arena, so the marmosets are coming in and out of the field of view. The marmosets are marked by shaving a certain segment of their tail. I'm labelling 9 body parts - ears, nose, hands, legs, tail base and tail end for each individual. I'm working with 300 labelled images to train 50,000 iterations of ResNet152 (after making sure that the loss plateaus). I'm facing a lot of issues with ID switching wherein the points are assigned to individuals A and B when only A is in the frame. It also happens often that ears, nose and hands are assigned to A and legs and tail are assigned to B. There are too many ID switching errors to correct them manually in refine_tracklets function. I also trained the network on the same data with ID=False, but this didn't help with the ID switching. Here is a link to the google drive folder with the detection videos - https://drive.google.com/drive/folders/1jGBq3F5WLbIjX1c8CcsC0J4JnGjcZwYf?usp=sharing. 

I am working with 10 different individuals and 4 maDLC projects for 4 cameras, so any increase in labelling effort will be quadrupled. Could you tell me how I can get good detections with consistent ID assignment for my videos?

Thank you so much for your help!

### Operating System

Windows 11

### DeepLabCut version

DLC 2.3.6

### DeepLabCut mode

multi animal

### Device type

GPU - NVIDIA RTX A2000

### Anything else?

![Cam10-11-15-18](https://github.com/DeepLabCut/DeepLabCut/assets/97757713/c47ef0c8-4e9c-4bfd-a5e0-16f792d2f42d)

Orientation of the cameras and experimental arena.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by jeylau** [comment]
Hello @vasudha-kulkarni, I'd encourage you to run the "keypoint assembly" part of the DLC workflow without relying on identity prediction (as it looks to me from your description that ID assignment is unreliable). To do so, run `deeplabcut.analyze_videos(..., auto_track=False)`, followed by the steps 10 and 11 of the maDLC tutorial https://deeplabcut.github.io/DeepLabCut/docs/tutorial.html:
`deeplabcut.convert_detections2tracklets(..., identity_only=False)` and `deeplabcut.stitch_tracklets`.
This way, points will be grouped together to form distinct animals using the "skeleton" and the "limbs" that the network learned (which if you've trained long enough should carry robust information), rather than relying on identity only.

**Comment by vasudha-kulkarni** [comment]
Hi Jessy,

Thank you for your reply. I have been working with `auto_track=False` and I had also tried an iteration with `identity=False` in the config.yaml file while training the network. Based on your suggestion, I re-analysed the video with `identity_only=False`, but the results weren't different. See the resulting video with filtered detections here - https://drive.google.com/file/d/11SkVmQzmfuEFN5ujSUCxDPl-v4_noZM4/view?usp=drive_link. 

I'm currently working with 300 training images because a pilot project (with different camera angle) had worked pretty well with only 100 training images. Do you think the ID switching might improve with more training data? If yes, how many frames would you suggest?

Thanks!

**Comment by jeylau** [comment]
Thanks for sharing the video! Perhaps the "edges" connecting pairs of body parts, learned by the network, are too weak (or, in other words, are not discriminative enough to robustly connect keypoints belonging to the same animal). Could you try to run `deeplabcut.utils.plot_edge_affinity_distributions(eval_pickle_path)`, passing in the location of the `_full.pickle` file normally found in the evaluation-results folder? The resulting plot (and the overlap between the 2 colored distributions) will help us diagnose what may be wrong.

**Comment by vasudha-kulkarni** [comment]
Hi Jessy,

Here is the resulting plot from the `deeplabcut.utils-plot_edge_affinity_distributions(_full.pickle)` from the iteration where I ran `identity=False`. Thanks!
![LL-maDLC-iter6-edge-affinity-distribution](https://github.com/DeepLabCut/DeepLabCut/assets/97757713/531d5fa7-9318-48d0-b293-6ddb59913f73)

**Comment by jeylau** [comment]
Thanks @vasudha-kulkarni! It does look like many pairwise connections do not allow to clearly distinguish an animal from the other (i.e., these with low scores and overlapped distributions). I believe the default `pafthreshold` is 0.1 in the `inference_cfg.yaml` file under `dlc-models/.../test`. Could you maybe retry with a lower threshold (0.05 or 0.01)? This will force DLC to consider more edges when grouping points together.

**Comment by vasudha-kulkarni** [comment]
Hello, thank you so much for your suggestion. Lowering the `pafthreshold` to 0.05 did reduce ID switching (see this video - https://drive.google.com/file/d/1t1QnNGyJIhHLh5MGIpuv57nznlLNH6Xw/view?usp=sharing). There are very few instances where head and limb body parts are assigned to different individuals (like in the image below), which was my initial problem. 
<img width="235" alt="image" src="https://github.com/DeepLabCut/DeepLabCut/assets/97757713/2fe175a0-31c2-46e6-8c4d-34a65e8e28cf">
But there were still ID switches when the animals move in and out of view, and a couple of times even when there's a single animal in view (at least 9 noticeable switches in 2 min video). Let me know if you have any other suggestions to improve this further.
Thanks again!
